### PR TITLE
Add graham scan algorithm

### PR DESCRIPTION
**Pull request by ulwlu**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: other/graham_scan.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    sorted_points.insert(0, (minx, miny))
+
+    # traversal from the lowest and the most left point in anti-clockwise direction
+    # if direction gets right, the previous point is not the convex hull.
+    class Direction(Enum):
+        left = 1
+        straight = 2
+        right = 3
+
+    def check_direction(
```
[the remark this answers is not in the document]
Comment (ulwlu) [reply other/graham_scan.py:79]: To reviewers:
This is an inner function

File: other/graham_scan.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            minidx = i
+        if y == miny:
+            if x < minx:
+                minx = x
+                minidx = i
+
+    # remove the lowest and the most left point from points for preparing for sort
+    points.pop(minidx)
+
+    def angle_comparer(p: list[int, int]) -> float:
```
[the remark this answers is not in the document]
Comment (ulwlu) [reply other/graham_scan.py]: To reviewers:
This is an inner function.

argument name is fixed here fdaacde

File: other/graham_scan.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        # sort the points accorgind to the angle from the lowest and the most left point
+        x = p[0]
+        y = p[1]
+        angle = degrees(atan2(y - miny, x - minx))
+        return angle
+
+    sorted_points = sorted(points, key=lambda p: angle_comparer(p))
```
[the remark this answers is not in the document]
Comment (ulwlu) [reply other/graham_scan.py]: argument name is fixed here fdaacde

**Comment by ulwlu** [comment]
To reviewers:

The test is failed at pre-commit because CI executes `black` for whole files in this repository, and it changes 814 files.

I did `black` with its newest version, and so CI does. It should be updated, or ignore this fail.

**Review by Rada8732 — APPROVED** [review]
``

**Comment by ulwlu** [comment]
@cclauss 
Hi, I'm sorry if I'm mentioning the wrong person, but you were the reviewer of my previous PR.

Tests look not falling anymore, but the label doesn't get removed. 
If this PR is not reviewed because of this label, I would be pleased if you could review when you have a time.

**Comment by cclauss** [comment]
Doesn't this algorithm belong in https://github.com/TheAlgorithms/Python/tree/master/sorts ?

**Comment by ulwlu** [comment]
@cclauss 
Hi. I'm not sure if this algorithm should belong to sorts, because this aims to **find the convex hull of a finite set of points**. It's kind of coordinate things rather than sorting.

**Comment by cclauss** [comment]
Putting algorithms in `other` makes them very hard for visitors to find.  Is there another category that makes sense?

**Review by cclauss — APPROVED** [review]

**Comment by ulwlu** [comment]
@cclauss 

I see. `searches` must be fitting. I will move this.

> How does this algorithm compare with master/divide_and_conquer/convex_hull.py

divide_and_conquer algorithm is also known well to solve convex hull problem only by O(n * logn) time complexity. However, it's approach is totally different from graham scan.

Graham scan doesn't use divide_and_conquer approach, but it focuses on sorting points by **angle**.
We'll look for all points at same time and trace the periphery with checking if the angle is towards to inner or outer.

**Comment by ulwlu** [comment]
Oh. Thank you for merging.

I wonder if I should create other PR to move this to `searches`, or shouldn't....

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/blob/master/divide_and_conquer/convex_hull.py has three different convex_hull algorithms and your file has a fourth.  What would be really cool would be to [benchmark](https://github.com/TheAlgorithms/Python/search?q=benchmark) the four algorithms on one or more common datasets to see the relative performance.

**Pull request merged**